openerp/addons/base/ir/ir_token.py

```diff
@@ -0,0 +1,103 @@
+        ('token_value_uniq', 'unique(value)', "Duplicated token")
+    ]
+
+    @api.model_cr
+    def _transient_clean_rows_older_than(self, seconds):
+        """Token have a minimum of 24 hours expiration before being cleaned up"""
+        seconds = max(seconds, 24 * 60 * 60)
+        return super(IrToken, self)._transient_clean_rows_older_than(seconds)
+
+    @api.multi
+    def _generate_token_url(self, route):
+        """ Returns a public URL using a custom route
+
+        :param route: URI in the form "/first/second/%s" with mandatory '%s' that
+            will be replaced by the token value.
+        """
+        self.ensure_one()
+
+        if "%s" not in route:
+            raise ValueError(_("Invalid route format %s, missing '%%s' to integrate token)") % route)
+
+        base_url = self.env['ir.config_parameter'].get_param('web.base.url')
+        return urljoin(base_url, route % self.value)
+
+    @api.model
+    def _find_valid_token(self, token, res_model=False, res_id=False):
+        """ Find a valid token based on record filters
+
+        A token is only valid 24 hours after its creation.
+
+        :param token: submitted token content
+        :param res_model: model name of the record linked to submitted token (optional)
+        :param res_id: id of the record linked to submitted token (optional)
+        """
+        yesterday = fields.Datetime.from_string(fields.Datetime.now()) - timedelta(days=1)
+        domain = [('value', '=', token), ('create_date', '>', fields.Datetime.to_string(yesterday))]
+        domain += res_model and [('res_model', '=', res_model)] or []
+        domain += res_id and [('res_id', '=', res_id)] or []
+        return self.sudo().search(domain, limit=1)
+
+    @api.model
+    def check_access_rights(self, operation, raise_exception=True):
+        """ Prevent anyone but administrator to access a token
+
+        Tokens may be linked to sensitive transactions so unrestricted access is
+        prevented.
+        """
+        if self.env.user._is_admin():
+            return True
+
+        if raise_exception:
+            raise AccessError(_("Sorry, you are not allowed to access this document."))
+        else:
+            return False
+
+    @api.multi
+    def check_access_rule(self, operation):
+        if self.env.user._is_admin():
+            return True
+
+        raise AccessError(_("Sorry, you are not allowed to access this document."))
+
+    @api.model
+    def _search(self, args, offset=0, limit=None, order=None, count=False, access_rights_uid=None):
+        """ Prevent anyone but administrator to find tokens """
+        if self.env.user._is_admin():
+            return super(IrToken, self)._search(args, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
+
+        return super(IrToken, self)._search(expression.FALSE_DOMAIN, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
```

Comment:
well actually, we don't really need to override `_search`, do we ? it's caught by `check_access_rights` anyway